python/valis/cache.py

```diff
@@ -0,0 +1,288 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Filename: main.py
+# Project: app
+# Author: José Sánchez-Gallego
+# Created: Monday, 9th December 2024
+# License: BSD 3-clause "New" or "Revised" License
+# Copyright (c) 2020 José Sánchez-Gallego
+# Last Modified: Monday, 9th December 2024
+# Modified By: José Sánchez-Gallego
+
+from __future__ import annotations
+
+import hashlib
+import json
+import logging
+from contextlib import asynccontextmanager
+from functools import wraps
+from inspect import Parameter, isawaitable, iscoroutinefunction
+from typing import (
+    TYPE_CHECKING,
+    Awaitable,
+    Callable,
+    List,
+    Optional,
+    ParamSpec,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast
+)
+
+from fastapi.dependencies.utils import (
+    get_typed_return_annotation,
+    get_typed_signature
+)
+from fastapi_cache import Backend, FastAPICache
+from fastapi_cache.backends.inmemory import InMemoryBackend
+from fastapi_cache.backends.redis import RedisBackend
+from fastapi_cache.decorator import _augment_signature, _locate_param
+from redis.asyncio.client import Redis
+from starlette.requests import Request
+from starlette.responses import Response
+from starlette.status import HTTP_304_NOT_MODIFIED
+
+from valis.settings import settings
+
+
+if TYPE_CHECKING:
+    from typing import AsyncIterator
+
+    from fastapi import FastAPI
+    from fastapi_cache.coder import Coder
+    from fastapi_cache.types import KeyBuilder
+
+
+__all__ = ['valis_cache', 'lifespan', 'valis_cache_key_builder']
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+logger = logging.getLogger("uvicorn.error")
+
+CACHE_TTL: float = 15_552_000 # 6 months
+
+
+@asynccontextmanager
+async def lifespan(_: FastAPI) -> AsyncIterator[None]:
+    backend = settings.cache_backend
+    if backend == 'in-memory':
+        logger.info('Using in-memory backend for caching')
+        FastAPICache.init(InMemoryBackend(),
+                          prefix="fastapi-cache",
+                          key_builder=valis_cache_key_builder)
```

Comment:
With the in-memory cache and the 6-month cache lifespan, I guess in principle someone's memory could fill up as they hit the routes?  Is that correct to say?  In practice since this is for development, I don't think it's a real issue.

Reply:
With in-memory the cache would clear when you stop running the app. I've added a note saying that this should only be used for development.